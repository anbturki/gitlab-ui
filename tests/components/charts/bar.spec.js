import { shallowMount } from '@vue/test-utils';
import BarChart from '../../../src/components/charts/bar/bar.vue';
import Chart from '../../../src/components/charts/chart/chart.vue';
import ChartTooltip from '../../../src/components/charts/tooltip/tooltip.vue';
import TooltipDefaultFormat from '../../../src/components/shared_components/charts/tooltip_default_format.vue';

const mockChartInstance = {
  getDom: () => {
    return {
      addEventListener: jest.fn(),
      removeEventListener: jest.fn(),
    };
  },
};

jest.mock('echarts', () => ({
  getInstanceByDom: () => mockChartInstance,
}));

const defaultChartProps = {
  xAxisTitle: 'Pushes per day',
  xAxisType: 'value',
  yAxisTitle: 'User',
  data: {
    Office: [
      [100, 'Jim'],
      [300, 'Pam'],
    ],
  },
};

describe('Bar chart component', () => {
  let wrapper;

  const findChart = () => wrapper.find(Chart);
  const findChartTooltip = () => wrapper.find(ChartTooltip);
  const getOptions = () => findChart().props('options');
  const emitChartCreated = () => findChart().vm.$emit('created', mockChartInstance);

  beforeEach(() => {
    wrapper = shallowMount(BarChart, {
      propsData: defaultChartProps,
      stubs: {
        'tooltip-default-format': TooltipDefaultFormat,
      },
    });
    emitChartCreated();

    return wrapper.vm.$nextTick();
  });

  afterEach(() => {
    wrapper.destroy();
  });

  it('emits "created" when onCreated is called', () => {
    wrapper.vm.onCreated(wrapper.vm.chart);

    expect(wrapper.emitted('created')).toBeTruthy();
  });

  describe('Tooltip', () => {
    const labelParams = {
      value: 'Jim',
      seriesName: 'Office',
      seriesData: [
        {
          seriesId: 'Office',
          data: [100, 'Jim'],
          value: [100, 'Jim'],
        },
      ],
    };

    it('calls custom method and returns title with y labels', () => {
      const getDefaultTooltipContentReturnVal = { yLabels: [], tooltipContent: {} };

      // stub the method
      wrapper.vm.getDefaultTooltipContent = jest
        .fn()
        .mockReturnValue(getDefaultTooltipContentReturnVal);

      // label formatter is triggered here which should eventually call
      // getDefaultTooltipContent
      getOptions().yAxis.axisPointer.label.formatter(labelParams);

      return wrapper.vm.$nextTick(() => {
        expect(wrapper.vm.getDefaultTooltipContent).toHaveBeenCalled();
      });
    });

    it('sets accurate tooltip title and content state', () => {
      getOptions().yAxis.axisPointer.label.formatter(labelParams);

      return wrapper.vm.$nextTick(() => {
        expect(wrapper.vm.tooltipTitle).toBe(labelParams.value);
        expect(wrapper.vm.tooltipContent[defaultChartProps.xAxisTitle].value).toBe(
          labelParams.seriesData[0].data[0]
        );
      });
    });

    it('render accurate tooltip title and content', () => {
      const expectedTooltipTitle = `${labelParams.value} (${defaultChartProps.yAxisTitle})`;

      getOptions().yAxis.axisPointer.label.formatter(labelParams);
      return wrapper.vm.$nextTick(() => {
        const tooltipTextContent = findChartTooltip().element.textContent;

        expect(tooltipTextContent).toContain(expectedTooltipTitle);
        expect(tooltipTextContent).toContain(defaultChartProps.xAxisTitle);
        expect(tooltipTextContent).toContain(labelParams.seriesData[0].data[0]);
      });
    });
  });
});
