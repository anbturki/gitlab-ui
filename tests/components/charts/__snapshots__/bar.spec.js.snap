// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Bar chart component when mounted should render chart with axis and series 1`] = `
Object {
  "grid": Object {
    "bottom": 44,
    "containLabel": true,
    "left": 22,
    "right": 32,
    "top": 16,
  },
  "legend": Object {
    "show": false,
  },
  "series": Array [
    Object {
      "barMaxWidth": "50%",
      "data": Array [
        Array [
          100,
          "Jim",
        ],
        Array [
          300,
          "Pam",
        ],
      ],
      "emphasis": Object {
        "itemStyle": Object {
          "color": "rgba(31, 120, 209, 0.4)",
        },
      },
      "itemStyle": Object {
        "barBorderColor": "#1f78d1",
        "barBorderWidth": 1,
        "color": "rgba(31, 120, 209, 0.2)",
      },
      "name": "Office",
      "stack": "chart",
      "type": "bar",
    },
  ],
  "xAxis": Object {
    "axisLabel": Object {
      "formatter": [Function],
    },
    "axisLine": Object {
      "show": false,
    },
    "name": "Pushes per day",
    "nameLocation": "center",
    "type": "value",
  },
  "yAxis": Object {
    "axisLabel": Object {
      "interval": 0,
    },
    "axisPointer": Object {
      "label": Object {
        "formatter": [Function],
      },
      "show": true,
      "type": "none",
    },
    "axisTick": Object {
      "show": true,
    },
    "boundaryGap": true,
    "name": "User",
    "nameGap": 60,
    "nameLocation": "center",
    "splitLine": Object {
      "show": false,
    },
    "type": "category",
  },
}
`;
